src/citrine/_utils/batcher.py

```diff
@@ -0,0 +1,164 @@
+            if isinstance(obj, HasParameters):
+                for attr in obj.parameters:
+                    if attr.template is not None:
+                        depends[obj].append(attr.template)
+            if isinstance(obj, HasConditions):
+                for attr in obj.conditions:
+                    if attr.template is not None:
+                        depends[obj].append(attr.template)
+            if isinstance(obj, HasProperties):
+                for attr in obj.properties:
+                    if attr.template is not None:
+                        depends[obj].append(attr.template)
+            if isinstance(obj, MaterialSpec):
+                for attr in obj.properties:
+                    if attr.property.template is not None:
+                        depends[obj].append(attr.property.template)
+                    for condition in attr.conditions:
+                        if condition.template is not None:
+                            depends[obj].append(condition.template)
+
+            # Index direct attribute template dependencies
+            if isinstance(obj, HasPropertyTemplates):
+                for attr in obj.properties:
+                    depends[obj].append(attr[0])
+            if isinstance(obj, HasConditionTemplates):
+                for attr in obj.conditions:
+                    depends[obj].append(attr[0])
+            if isinstance(obj, HasParameterTemplates):
+                for attr in obj.parameters:
+                    depends[obj].append(attr[0])
+
+            if isinstance(obj, HasTemplate):
+                if obj.template is not None:
+                    depends[obj].append(obj.template)
+            if isinstance(obj, (IngredientRun, MaterialRun, MeasurementRun, ProcessRun)):
+                if obj.spec is not None:
+                    depends[obj].append(obj.spec)
+            if isinstance(obj, (IngredientRun, IngredientSpec, MeasurementRun)):
+                if obj.material is not None:
+                    depends[obj].append(obj.material)
+            if isinstance(obj, (IngredientRun, IngredientSpec, MaterialRun, MaterialSpec)):
+                if obj.process is not None:
+                    depends[obj].append(obj.process)
+
+            by_type[obj.typ].append(obj)
+
+        # Deep dependencies w/ objects only, build inverse index
+        # This takes a second loop because we need to build up all derefs first
+        supported_by = defaultdict(list)
+        type_groups = sorted(list(by_type.values()), key=lambda x: writable_sort_order(x[0]))
+        for type_group in type_groups:
+            for obj in type_group:
+                local_set = {derefs[x] for x in depends[obj] if x in derefs}
+                full_set = set(local_set)
+                if len(full_set) > batch_size:
+                    raise ValueError("Dependency chain larger than batch size: "
+                                     f"{obj.name} > {batch_size}")
```

Comment:
Can we improve this error message?

To something like "Object {obj.name} has greater than {batch_size} dependencies."